From 1eb2af928ada1a00f26d7379e00c3c3f7a9749f4 Mon Sep 17 00:00:00 2001
From: Kevin McNee <mcnee@chromium.org>
Date: Thu, 14 Jul 2022 16:15:42 +0000
Subject: [PATCH 086/110] M102: Use weak ptr for webview JavaScriptDialogHelper
 callback

Use weak ptr for webview JavaScriptDialogHelper callback

This can be called asynchronously, potentially after the associated
WebViewGuest is destroyed.

(cherry picked from commit 1c09b9292dba7dfdc28b9bd09c61e3a0faf7b302)

Bug: 1336266
Change-Id: I8a4ec5ab124a9d5ca2ad45b1915666c8b7c98f79
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3715049
Reviewed-by: James Maclean <wjmaclean@chromium.org>
Auto-Submit: Kevin McNee <mcnee@chromium.org>
Commit-Queue: James Maclean <wjmaclean@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1015960}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3755322
Commit-Queue: Kevin McNee <mcnee@chromium.org>
Cr-Commit-Position: refs/branch-heads/5005@{#1245}
Cr-Branched-From: 5b4d9450fee01f821b6400e947b3839727643a71-refs/heads/main@{#992738}
---
 .../browser/guest_view/web_view/javascript_dialog_helper.cc    | 2 +-
 .../browser/guest_view/web_view/javascript_dialog_helper.h     | 3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/extensions/browser/guest_view/web_view/javascript_dialog_helper.cc b/extensions/browser/guest_view/web_view/javascript_dialog_helper.cc
index 677f609d3fd0..055a909d599b 100644
--- a/extensions/browser/guest_view/web_view/javascript_dialog_helper.cc
+++ b/extensions/browser/guest_view/web_view/javascript_dialog_helper.cc
@@ -66,7 +66,7 @@ void JavaScriptDialogHelper::RunJavaScriptDialog(
   web_view_permission_helper->RequestPermission(
       WEB_VIEW_PERMISSION_TYPE_JAVASCRIPT_DIALOG, request_info,
       base::BindOnce(&JavaScriptDialogHelper::OnPermissionResponse,
-                     base::Unretained(this), std::move(callback)),
+                     weak_factory_.GetWeakPtr(), std::move(callback)),
       false /* allowed_by_default */);
 }
 
diff --git a/extensions/browser/guest_view/web_view/javascript_dialog_helper.h b/extensions/browser/guest_view/web_view/javascript_dialog_helper.h
index 53d49b7dc5c5..ccd039d86d8e 100644
--- a/extensions/browser/guest_view/web_view/javascript_dialog_helper.h
+++ b/extensions/browser/guest_view/web_view/javascript_dialog_helper.h
@@ -6,6 +6,7 @@
 #define EXTENSIONS_BROWSER_GUEST_VIEW_WEB_VIEW_JAVASCRIPT_DIALOG_HELPER_H_
 
 #include "base/memory/raw_ptr.h"
+#include "base/memory/weak_ptr.h"
 #include "content/public/browser/javascript_dialog_manager.h"
 
 namespace extensions {
@@ -46,6 +47,8 @@ class JavaScriptDialogHelper : public content::JavaScriptDialogManager {
 
   // Pointer to the webview that is being helped.
   const raw_ptr<WebViewGuest> web_view_guest_;
+
+  base::WeakPtrFactory<JavaScriptDialogHelper> weak_factory_{this};
 };
 
 }  // namespace extensions
-- 
2.38.1

